### ISSUE TITLE
Kubelet removes non-Kubernetes images.

### ISSUE THREAD

**Issue by akunszt** [kubernetes/kubernetes#109936]
### What happened?

We also hit by this issue. We're using Flatcar Linux which is quite barebone so we are using a few containers to augment it. Unfortunately kubelet removes the images when they are not in active use. For example we a have a cronjob which uses an image we pulled and specially tagged. That images isn't used by any pods at all but kubelet choose to remove it under disk pressure. It is very annoying.

It would be nice to have a setting which says that "you can touch images only if the tag is starting with this prefix" or similar.

Even if you are using another Linux distribution then you might want to install something which isn't available in the distribution itself. Using a container is a very likely way to go and you don't expect that something which never belonged to the Kubernetes cluster will be removed by kubelet. It's very unintuitive.

### What did you expect to happen?

Kubelet won't remove images which aren't belongs to it.

### How can we reproduce it (as minimally and precisely as possible)?

* Pull some images from the command line.
* Fill up the local disk to create a DiskPressure "situation".
* Wait until kubelet removes the images - by the documentation it does in every five minutes.

### Anything else we need to know?

I created this new ticket because the old one (#103255) I found gets closed by k8s-ci-robot.

We are still using Kubernetes 1.21 as because of the faulty etcd-client lib used in 1.22 and 1.23 we cannot upgrade our clusters. The fix appeared in 1.24 but never backported to 1.22 or 1.23 which left us no possible upgrade paths.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Kubernetes cluster running on EC2 instances in AWS.
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=3033.2.3
VERSION_ID=3033.2.3
BUILD_ID=2022-03-02-2032
PRETTY_NAME="Flatcar Container Linux by Kinvolk 3033.2.3 (Oklo)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar-linux.org/"
BUG_REPORT_URL="https://issues.flatcar-linux.org"
FLATCAR_BOARD="amd64-usr"
$ uname -a
Linux ip-10-8-74-48.ec2.internal 5.10.102-flatcar #1 SMP Wed Mar 2 19:40:13 -00 2022 x86_64 AMD EPYC 7571 AuthenticAMD GNU/Linux
```

</details>

### Install tools

<details>
None.
</details>

### Container runtime (CRI) and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

**Comment by BenTheElder** [comment]
Currently kubelet only has "don't touch pause".

A feature like this will need an enhancement proposal. we've started one before but haven't had time to go through with it https://github.com/kubernetes/enhancements/pull/1717 / https://github.com/kubernetes/enhancements/pull/1007

**Comment by akunszt** [comment]
I see both proposals are closed, so this won't be happening I assume. Am I right?
Is there anything we can do to make this happen?

**Comment by dims** [comment]
@akunszt very little probability. you can join the sig-node meeting and ask there.

**Comment by harshanarayana** [comment]
@BenTheElder  @dims  I would be happy to help in any way I can with this if that is all right with you. In our product we have a local docker registry that runs as a Pod in the local k8s cluster itself. Which like you can already see is a cyclic dependency. Though it is not the best thing, works great for us in AirGap solutions. We were bit by this uncontrolled GC in production a bunch of times. 

The GC manager ended up cleaning up the images required to run the registry itself as part of the GC and as a side effect it took part of the cluster down with it. 

Since we are managing a custom k8s ourself internally, we went in an put an exception list that we let the end user configure via the kubelet config. The exception file is a JSON file you can specify with support to excluding images via tag/image ID. We customised the Image GC manager to honour this and it has been helping us well so far. Would be more than happy to help get that in the upstream if the community agrees to it.

**Comment by dims** [comment]
@harshanarayana 2 ways to do this ... easy way right now would be to just send what you have now as a draft PR to kick start the discussion (esp to determine if a KEP is needed).

**Comment by harshanarayana** [comment]
@dims Great. Let me get that done by End of next week. Mind assigning this to me ?

**Comment by BenTheElder** [comment]
xref: https://github.com/containerd/containerd/issues/6930

**Comment by harshanarayana** [comment]
@BenTheElder  /Ack. I also remember d3ae0a381ac8d90d8a532bd502e685a2641ea9db was merged a while ago. To enable exempting the pinned images from being GC'ed.

**Comment by BenTheElder** [comment]
Though this comment suggests it's only a a recommendation to kubelet not a requirement https://github.com/kubernetes/kubernetes/commit/d3ae0a381ac8d90d8a532bd502e685a2641ea9db#diff-583fcb2b2b9fed64e8ac45392565a924918b4a64a7d4b66df0d4bccd298e8769R1259

**Comment by harshanarayana** [comment]
Indeed. Also, the current implementation like you shared above, is only addressing the sandbox images. (Until the generalisation discussion is concluded on that thread) What would be the way one would indicate that this image should be pinned  on the K8s side when the pod is being created ? Annotate it in the spec ?

**Comment by BenTheElder** [comment]
> What would be the way one would indicate that this image should be pinned on the K8s side when the pod is being created ? Annotate it in the spec ?

I don't think pods should be able to configure this at all, I think this is something for cluster administrators to configure.
Letting pods configure it is a potential DOS issue.

**Comment by harshanarayana** [comment]
/Ack @BenTheElder  @dims 

I have a draft PR opened with the changes that we use internally to have an exemption list. PTAL and let me know if we should take this via a KEP.

**Comment by BenTheElder** [comment]
IMHO: At this point this is the responsibility of the CRI implementation. Kubelet makes image management calls against CRI, and the CRI implementation should be scoping them to Kubernetes (e.g. in containerd they are in the k8s.io namespace which is reserved for Kubernetes / kubelet).

dockershim no longer exists in Kubelet and cri-dockershim could conceivably handle this. containerd at least already does.
